daemon/metrics.go

```diff
@@ -0,0 +1,84 @@
+package daemon
+
+import (
+	"github.com/docker/docker/container"
+	"github.com/docker/docker/dockerversion"
+	"github.com/docker/engine-api/types"
+	"github.com/docker/go-metrics"
+)
+
+var (
+	containerActions   metrics.LabeledTimer
+	imageActions       metrics.LabeledTimer
+	containerCounter   metrics.LabeledCounter
+	daemonActions      metrics.LabeledTimer
+	networkActions     metrics.LabeledTimer
+	engineVersion      metrics.LabeledGauge
+	healthCheckCounter metrics.Counter
+)
+
+func init() {
+	ns := metrics.NewNamespace("engine", "daemon", metrics.Labels{
+		"version": dockerversion.Version,
+		"commit":  dockerversion.GitCommit,
+	})
+	containerActions = ns.NewLabeledTimer("container_actions", "The number of seconds it takes to process each container action", "action")
+	daemonActions = ns.NewLabeledTimer("daemon_actions", "The number of seconds it takes to process each daemon action", "action")
+	networkActions = ns.NewLabeledTimer("network_actions", "The number of seconds it takes to process each network action", "action")
+	engineVersion = ns.NewLabeledGauge("version", "The version and commit information for the engine process", metrics.Unit("info"),
+		"architecture",
+		"cpus",
+		"graph_driver",
+		"kernel",
+		"memory",
+		"os",
+	)
+	containerCounter = ns.NewLabeledCounter("container", "The total number of containers processed", "action")
+	healthCheckCounter = ns.NewCounter("health_check", "The total number of health checks executed")
+	imageActions = ns.NewLabeledTimer("image_actions", "The number of seconds it takes to process each image action", "action")
+	metrics.Register(ns)
+}
+
+func (daemon *Daemon) startMetrics(c *container.Container) {
+	ns := metrics.NewNamespace("container", "stat", metrics.Labels{
+		"id":    c.ID,
+		"image": c.Config.Image,
```

Comment:
@jhorwit2 There is a trick in prometheus where you can join a metric based on some parameter. In this case, we increment a static counter that contains the meta data (labels, name, etc.) and container id. The rest of the metrics, we simply emit with container id. When you want to know about labels and counts, you join the metadata with the container id. This approach is generally applicable most data models.

This allows us to reduce the volume of data being emitted per container while still binding rich metadata to the metrics.